hack/tools/conversion-verifier/main.go

```diff
@@ -0,0 +1,196 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"go/types"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+var (
+	storageVersion   = markers.Must(markers.MakeDefinition("kubebuilder:storageversion", markers.DescribesType, crdmarkers.StorageVersion{}))
+	nolintConversion = markers.Must(markers.MakeDefinition("kubebuilder:conversion:nolint", markers.DescribesType, false))
+
+	apiPackages         = map[string]map[string]*loader.Package{}
+	storageVersionTypes = map[string]*storageVersionType{}
+)
+
+type versionType struct {
+	base string
+	pkg  *loader.Package
+	typ  *markers.TypeInfo
+}
+
+func (v *versionType) IsConvertible() bool {
+	return hasMethod(v.pkg, v.typ, "ConvertFrom") && hasMethod(v.pkg, v.typ, "ConvertTo")
+}
+
+type storageVersionType struct {
+	*versionType
+
+	otherDecls map[string]*versionType
+}
+
+func (s *storageVersionType) IsHub() bool {
+	return hasMethod(s.pkg, s.typ, "Hub")
+}
+
+func main() {
+	var result error
+
+	// Define the marker collector.
+	col := &markers.Collector{
+		Registry: &markers.Registry{},
+	}
+	// Register the markers.
+	if err := col.Registry.Register(storageVersion); err != nil {
+		klog.Fatal(err)
+	}
+	if err := col.Registry.Register(nolintConversion); err != nil {
+		klog.Fatal(err)
+	}
+
+	// Load all packages.
+	packages, err := loader.LoadRoots("./...")
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	// First loop through all the packages
+	// and find the ones that have a storage version assigned.
+	for i := range packages {
+		root := packages[i]
+
+		base := basePackage(root.PkgPath)
+		if !strings.HasSuffix(base, "/api") {
+			// We're only interested in api folders, exclude everything
+			// else that's not an API package.
+			continue
+		}
+
+		// Store the packages.
+		if _, ok := apiPackages[base]; !ok {
+			apiPackages[base] = make(map[string]*loader.Package)
+		}
+		pkgs := apiPackages[base]
+		pkgs[root.PkgPath] = root
+
+		// Populate type information and loop through every type.
+		root.NeedTypesInfo()
+		if err := markers.EachType(col, root, func(info *markers.TypeInfo) {
+			// Check if the type is registered with storage version.
+			if m := info.Markers.Get(storageVersion.Name); m == nil {
+				return
+			}
+			if _, ok := storageVersionTypes[info.Name]; ok {
+				result = multierror.Append(result,
```

Comment:
what's benefit of introducing a dep into a new external package vs `var errors []error and apimachinery NewAggregate`?

Reply:
Nicer output

Reply:
Example:

```
F0923 08:56:38.548784   87515 main.go:214] 6 errors occurred:
        * type "cluster.x-k8s.io/MachinePoolList" in package "sigs.k8s.io/cluster-api/exp/api/v1beta1" marked as storage version but it's not convertible, missing Hub() method
        * type "cluster.x-k8s.io/MachinePoolList" in package "sigs.k8s.io/cluster-api/exp/api/v1alpha4" it's not convertible, missing ConvertFrom() and ConvertTo() methods
        * type "cluster.x-k8s.io/MachinePoolList" in package "sigs.k8s.io/cluster-api/exp/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
        * type "cluster.x-k8s.io/MachinePool" in package "sigs.k8s.io/cluster-api/exp/api/v1beta1" marked as storage version but it's not convertible, missing Hub() method
        * type "cluster.x-k8s.io/MachinePool" in package "sigs.k8s.io/cluster-api/exp/api/v1alpha4" it's not convertible, missing ConvertFrom() and ConvertTo() methods
        * type "cluster.x-k8s.io/MachinePool" in package "sigs.k8s.io/cluster-api/exp/api/v1alpha3" it's not convertible, missing ConvertFrom() and ConvertTo() methods
```